xtail.1
.\" $Id: xtail.1,v 2.2 1996/06/25 00:31:15 chip Exp $
.TH XTAIL 1L
.SH NAME
xtail - Watch the growth of files.
.SH SYNTAX
.B xtail
entry ...
.SH DESCRIPTION
.I Xtail
monitors one or more files, and displays all data written to a file
since command invocation.  It is very useful for monitoring multiple
logfiles simultaneously.
.P
If an
.I entry
given on the command line is a directory, all files in that directory
will be monitored, including those created after the
.I xtail
invocation.  If an
.I entry
given on the command line doesn't exist,
.I xtail
will watch for it and monitor it once created.  When switching files in
the display, a banner showing the pathname of the file is printed.
.P
An interrupt character (usually CTRL/C or DEL) will display a list of the
most recently modified files being watched.  Send a quit signal
(usually CTRL/backslash) to stop
.IR xtail .
.SH SEE ALSO
tail(1)
.SH NOTES
.I Xtail
may be easily confused.  For example, if a file is renamed,
.I xtail
may or may not continue to monitor it.  If you ask it to monitor a file
multiple times, it probably will.  If you misspell a filename,
.I xtail
will treat it as a nonexistent entry and happily wait for its creation.
.P
My favorite use is "xtail /usr/spool/uucp/.Log/*".
.SH AUTHOR
Chip Rosenthal <chip@unicom.com>


debian/changelog
xtail (2.1-7) unstable; urgency=medium

  * QA upload.
  * Using new DH level format. Consequently:
      - debian/compat: removed.
      - debian/control: changed from 'debhelper' to 'debhelper-compat' in
        Build-Depends field and bumped level to 12.
  * debian/control:
      - Added 'Rules-Requires-Root: no' to source stanza.
      - Bumped Standards-Version to 4.5.0.
      - Created VCS fields.
      - Reorganized the long description.
  * debian/copyright:
      - Reverted packaging licensing to BSD-3-Clause. (Closes: #956140)
      - Updated packaging copyright years.
  * debian/patches: added a numeric prefix for all patches.
  * debian/rules:
      - Added DEB_BUILD_MAINT_OPTIONS to improve the GCC hardening.
      - Removed trash.
  * debian/salsa-ci.yml: added to provide CI tests for Salsa.
  * debian/tests/control: created to perform a trivial CI test.
  * debian/watch: bumped version to 4.
  * debian/xtail.dirs: renamed to debian/dirs.

 -- Joao Eriberto Mota Filho <eriberto@debian.org>  Tue, 07 Apr 2020 00:39:15 -0300

xtail (2.1-6) unstable; urgency=medium

  * QA upload.
  * Source format 3.0 (quilt)
  * Bumped compat level to 10
  * Converted d/rules to dh format
  * Add ISO Date patch from Jari Aalto <jari.aalto@cante.net> (Closes: #562903)
  * Set maintainer to QA Group
  * Added homepage to d/control
  * Converted d/copyright to machine-readable format
  * Made URLs use HTTPS where possible
  * Standards version 4.0.0

 -- David William Richmond Davies-Jones <david@exultantmonkey.co.uk>  Wed, 19 Jul 2017 15:32:38 +0100

xtail (2.1-5.1) unstable; urgency=medium

  * Non-maintainer upload.
  * Update DH level to 9 to avoid a FTBFS. (Closes: #800269)
  * debian/compat: created.
  * debian/control:
      - Added the ${misc:Depends} variable to provide the
        right install dependencies.
      - Bumped Standards-Version to 3.9.6.
  * debian/rules: disabled the DH_COMPAT line.
  * debian/watch: updated. (Closes: #551417)

 -- Joao Eriberto Mota Filho <eriberto@debian.org>  Thu, 10 Dec 2015 17:54:42 -0200

xtail (2.1-5) unstable; urgency=low

  * Note in debian/control that this isn't an X11 program (closes:
    #426350).

 -- Roderick Schertler <roderick@argon.org>  Tue, 24 Jun 2008 11:25:57 -0400

xtail (2.1-4) unstable; urgency=low

  * Add do-nothing binary/indep target (closes: #395763).
  * Oops, files in directories given on the command line were output in
    full due to a bug in code I'd introduced (closes: #368897)

 -- Roderick Schertler <roderick@argon.org>  Sat, 18 Nov 2006 21:39:17 -0500

xtail (2.1-3) unstable; urgency=low

  * Drop /usr/doc symlink by rebuilding with newer debhelper (closes:
    #359611).
  * Note in the copyright file that that Chip Rosental really is the
    author (closes: #302920).
  * xtail.1:  Change embedded section from "1L" to "1".

 -- Roderick Schertler <roderick@argon.org>  Thu,  6 Apr 2006 10:25:06 -0400

xtail (2.1-2) unstable; urgency=low

  * Handle files which are renamed properly.

 -- Roderick Schertler <roderick@argon.org>  Fri,  7 Dec 2001 09:16:54 -0500

xtail (2.1-1) unstable; urgency=low

  * Initial version (closes: #111309).

 -- Roderick Schertler <roderick@argon.org>  Thu, 13 Sep 2001 17:23:10 -0400

$Id: changelog,v 1.5 2008-06-24 15:26:26 roderick Exp $


xtail.c
/* $Id: xtail.c,v 2.5 2000/06/04 09:09:03 chip Exp $ */

#include "config.h"
#include <sys/types.h>
#include <sys/stat.h>
#include <stdio.h>
#include <signal.h>
#ifdef HAVE_UNISTD_H
# include <unistd.h>
#endif
#ifdef HAVE_STDLIB_H
# include <stdlib.h>
#endif
#define  INTERN
#include "xtail.h"

int sigcaught = 0;
RETSIGTYPE sigcatcher(int sig);


int main(int argc, char *argv[])
{
    int open_files_only, already_open, iteration, i;
    struct entry_descrip *entryp;
    struct stat sbuf;

    /* 
     * Initialize.
     */
    Progname = basename(argv[0]);
    List_file = new_entry_list(ENTRY_LIST_CHUNK);
    List_dir = new_entry_list(ENTRY_LIST_CHUNK);
    List_zap = new_entry_list(ENTRY_LIST_CHUNK);
    Sorted = FALSE;
    Reset_status = FALSE;
    Debug = FALSE;
    sigcatcher(0);


    /*
     * Place all of the entries onto lists.
     */
    for (i = 1 ; i < argc ; ++i)  {

	if (i == 1 && strcmp(argv[i], "-D") == 0) {
	    Debug = TRUE;
	    continue;
	}

	/*
	 * Temporarily throw this entry onto the end of the zapped list.
	 */
	entryp = new_entry(List_zap, argv[i]);

	/*
	 * Stat the file and get it to its proper place.
	 */
	switch (stat_entry(List_zap, last_entry(List_zap), &sbuf)) {

	case ENTRY_FILE:		/* move entry to file list	*/
	    move_entry(List_file, List_zap, last_entry(List_zap));
	    entryp->size = sbuf.st_size;
	    entryp->mtime = sbuf.st_mtime;
	    break;

	case ENTRY_DIR:			/* move entry to dir list	*/
	    move_entry(List_dir, List_zap, last_entry(List_zap));
	    entryp->size = sbuf.st_size;
	    entryp->mtime = sbuf.st_mtime;
	    if (scan_directory(entryp->name) != 0) {
		message(MSSG_OPEN, entryp);
		rmv_entry(List_dir, last_entry(List_dir));
	    }
	    break;

	case ENTRY_ZAP:			/* keep entry on zap list	*/
	    break;

	case ENTRY_SPECIAL:		/* entry is a special file	*/
	    message(MSSG_NOTAFIL, entryp);
	    rmv_entry(List_zap, last_entry(List_zap));
	    break;

	default:			/* stat error			*/
	    message(MSSG_STAT, entryp);
	    rmv_entry(List_zap, last_entry(List_zap));
	    break;

	}

    }

    /*
     * Make sure we are watching something reasonable.
     */
    if (List_file->num_entries == 0) {
	if (List_dir->num_entries == 0 && List_zap->num_entries == 0) {
	    (void) fprintf(stderr, "%s: no valid entries specified\n", Progname);
	    (void) exit(1);
	}
	(void) puts("\n*** warning - no files are being watched ***");
    }


    /*
     * From this point on we want to reset the status of an entry any
     * time we move it around to another list.
     */
    Reset_status = TRUE;


    /*
     * Force a check of everything first time through the loop.
     */
    iteration = CHECK_COUNT;


    /* 
     * Loop forever.
     */
    for (;;) {

	/*
	 * Once every CHECK_COUNT iterations check everything.
	 * All other times only look at the opened files.
	 */
	open_files_only = (++iteration < CHECK_COUNT);
	if (!open_files_only)
	    iteration = 0;


	/*
	 * Make sure that the most recently modified files are open.
	 */
	if (!Sorted)
	    fixup_open_files();


	/*
	 * Display what we are watching if a SIGINT was caught.
	 */
	if (sigcaught) {
	    show_status();
	    sigcatcher(0);
	}


	/*
	 * Go through all of the files looking for changes.
	 */
	Dprintf(stderr, ">>> checking files list (%s)\n",
	    (open_files_only ? "open files only" : "all files"));
	for (i = 0 ; i < List_file->num_entries ; ++i) {

	    entryp = List_file->list[i];
	    already_open = (entryp->fd > 0) ;

	    /*
	     * Ignore closed files except every CHECK_COUNT iterations.
	     */
	    if (!already_open && open_files_only)
		continue;

	    /*
	     * Get the status of this file.
	     */
	    switch (stat_entry(List_file, i, &sbuf)) {
	    case ENTRY_FILE:		/* got status OK		*/
		break;
	    case ENTRY_DIR:		/* huh??? it's now a dir	*/
		move_entry(List_dir, List_file, i--);
		continue;
	    case ENTRY_ZAP:		/* entry has been deleted	*/
		message(MSSG_ZAPPED, entryp);
		move_entry(List_zap, List_file, i--);
		continue;
	    case ENTRY_SPECIAL:		/* entry is a special file	*/
		message(MSSG_NOTAFIL, entryp);
		rmv_entry(List_file, i--);
		continue;
	    default:			/* stat error			*/
		message(MSSG_STAT, entryp);
		rmv_entry(List_file, i--);
		continue;
	    }


	    /*
	     * See if an opened file has been deleted.
	     */
	    if (already_open && sbuf.st_nlink == 0) {
		message(MSSG_ZAPPED, entryp);
		move_entry(List_zap, List_file, i--);
		continue;
	    }

	    /*
	     * If nothing has changed then continue on.
	     */
	    if (entryp->size==sbuf.st_size && entryp->mtime==sbuf.st_mtime)
		continue;

	    /*
	     * If the file isn't already open, then do so.
	     *   Note -- it is important that we call "fixup_open_files()"
	     *   at the end of the loop to make sure too many files don't
	     *   stay opened.
	     */
	    if (!already_open && open_entry(List_file, i) != 0) {
		--i;
		continue;
	    }

	    /*
	     * See if the file has been truncated.
	     */
	    if (sbuf.st_size < entryp->size) {
		message(MSSG_TRUNC, entryp);
		entryp->size = 0;
	    }

	    /*
	     * Seek to where the changes begin.
	     */
	    if (lseek(entryp->fd, entryp->size, 0) < 0) {
		message(MSSG_SEEK, entryp);
		rmv_entry(List_file, i--);
		continue;
	    }

	    /*
	     * Dump the recently added info.
	     */
	    {
		int nb;
    		static char buf[BUFSIZ];
		message(MSSG_BANNER, entryp);
		while ((nb = read(entryp->fd, buf, sizeof(buf))) > 0) {
		    (void) fwrite(buf, sizeof(char), (unsigned) nb, stdout);
		    entryp->size += nb;
		}
		if (nb < 0) {
		    message(MSSG_READ, entryp);
		    rmv_entry(List_file, i--);
		    continue;
		}
	    }

	    /*
	     * Update the modification time.
	     */
	    entryp->mtime = sbuf.st_mtime;

	    /*
	     * Since we've changed the mtime, the list might no longer be
	     * sorted.  However if this entry is already at the top of the
	     * list then it's OK.
	     */
	    if (i != 0)
		Sorted = FALSE;

	    /*
	     * If we've just opened the file then force a resort now to
	     * prevent too many files from being opened.
	     */
	    if (!already_open)
		fixup_open_files();

	}


	/*
	 * Go through list of nonexistent entries to see if any have appeared.
	 *   This is done only once every CHECK_COUNT iterations.
	 */
	if (!open_files_only) {
	    Dprintf(stderr, ">>> checking zapped list\n");
	    for (i = 0 ; i < List_zap->num_entries ; ++i) {
		entryp = List_zap->list[i];
		switch (stat_entry(List_zap, i, &sbuf)) {
		case ENTRY_FILE:	/* entry has appeared as a file	*/
		    message(MSSG_CREATED, entryp);
		    move_entry(List_file, List_zap, i--);
		    break;
		case ENTRY_DIR:		/* entry has appeared as a dir	*/
		    message(MSSG_CREATED, entryp);
		    move_entry(List_dir, List_zap, i--);
		    break;
		case ENTRY_ZAP:		/* entry still doesn't exist	*/
		    break;
		case ENTRY_SPECIAL:	/* entry is a special file	*/
		    message(MSSG_NOTAFIL, entryp);
	    	    rmv_entry(List_zap, i--);
		    break;
		default:		/* error - entry removed	*/
	    	    message(MSSG_STAT, entryp);
	    	    rmv_entry(List_zap, i--);
		    break;
		}
	    }
	}


	/*
	 * Go through the list of dirs to see if any new files were created.
	 *   This is done only once every CHECK_COUNT iterations.
	 */
	if (!open_files_only) {
	    Dprintf(stderr, ">>> checking directory list\n");
	    for (i = 0 ; !open_files_only && i < List_dir->num_entries ; ++i) {
		entryp = List_dir->list[i];
		switch (stat_entry(List_dir, i, &sbuf)) {
		case ENTRY_DIR:		/* got status OK		*/
		    break;
		case ENTRY_FILE:	/* huh??? it's now a reg file	*/
		    move_entry(List_file, List_dir, i--);
		    continue;
		case ENTRY_ZAP:		/* entry has been deleted	*/
	    	    message(MSSG_ZAPPED, entryp);
	    	    move_entry(List_zap, List_dir, i--);
		    continue;
		case ENTRY_SPECIAL:	/* entry is a special file	*/
		    message(MSSG_NOTAFIL, entryp);
		    rmv_entry(List_dir, i--);
		    continue;
		default:		/* stat error			*/
		    message(MSSG_STAT, entryp);
		    rmv_entry(List_dir, i--);
		    continue;
		}
		if (entryp->mtime == sbuf.st_mtime)
		    continue;
		if (scan_directory(entryp->name) != 0) {
		    message(MSSG_OPEN, entryp);
		    rmv_entry(List_dir, i--);
		}
		entryp->mtime = sbuf.st_mtime;
	    }
	}


	/*
	 * End of checking loop.
	 */
	{
	    extern unsigned sleep();
	    (void) fflush(stdout);
	    (void) sleep(SLEEP_TIME);
	}

    }

    /*NOTREACHED*/

}


RETSIGTYPE sigcatcher(int sig)
{
    if (sig == SIGQUIT)
	(void) exit(0);
    sigcaught = sig;
#ifdef STATUS_ENAB
    (void) signal(SIGINT, sigcatcher);
    (void) signal(SIGQUIT, sigcatcher);
#endif
}

